### ISSUE TITLE
getSharedPreferences on UIThread

### ISSUE THREAD

**Issue by shaniamjad** [OneSignal/OneSignal-Android-SDK#299]
Hello Guys,

I am using OneSignal Android SDK version 3.5.8 in my game which is LIVE on Google play store for sometime. I have been receiving Lots of ANR reports from Google dashboard related to Illegal I/O operations from disk. 
After doing some investigation using StrictMode enabled i figured that OneSignal SDK gives error on initialisation, where is calls getGcmPreferences and tries to get sharedPreferences data inside this.
This function is being called from UIThread and since it is doing an I/O operation so it may leads to an ANR. So kindly help to fix this issue. Thanks
I am pasting full stack below so it can better be addressed.

```Java
D/StrictMode: StrictMode policy violation; ~duration=130 ms: android.os.StrictMode$StrictModeDiskReadViolation: policy=65567 violation=2
 at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1293)
 at java.io.UnixFileSystem.checkAccess(UnixFileSystem.java:249)
 at java.io.File.exists(File.java:780)
 at android.app.ContextImpl.ensurePrivateDirExists(ContextImpl.java:541)
 at android.app.ContextImpl.getPreferencesDir(ContextImpl.java:497)
 at android.app.ContextImpl.getSharedPreferencesPath(ContextImpl.java:656)
 at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:374)
 at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:171)
 at com.onesignal.OneSignal.getGcmPreferences(OneSignal.java:1672)
 at com.onesignal.OneSignal.saveFilterOtherGCMReceivers(OneSignal.java:1270)
 at com.onesignal.OneSignal.init(OneSignal.java:370)
 at com.onesignal.OneSignal.init(OneSignal.java:326)
 at com.onesignal.OneSignal.access$000(OneSignal.java:67)
 at com.onesignal.OneSignal$Builder.init(OneSignal.java:165)
 at com.modules.common.PushNotificationManagerAndroid.innitialiseSDK(PushNotificationManagerAndroid.java:22)
 at com.modules.common.ModuleCommon.onCreate(ModuleCommon.java:37)
 at org.cocos2dx.cpp.AppActivity.onCreate(AppActivity.java:84)
 at android.app.Activity.performCreate(Activity.java:6912)
 at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1126)
 at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2900)
 at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3008)
 at android.app.ActivityThread.-wrap14(ActivityThread.java)
 at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1650)
 at android.os.Handler.dispatchMessage(Handler.java:102)
 at android.os.Looper.loop(Looper.java:154)
 at android.app.ActivityThread.main(ActivityThread.java:6688)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1468)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1358)
```

**Comment by jkasten2** [comment]
@shaniamjad Could you attach the ANR as well to confirm the issue?

We will be moving all I/O operations off the main thread and/or do so in a non-main thread blocking way. In 3.6.0 we changed all `SharedPreferences` `commit` calls to `apply` which writes in a non-blocking way. We still need to move reads off the main thread but updating to 3.6.X should lower the occurrences of this specific ANR.

**Comment by shaniamjad** [comment]
I dont have an ANR which directly gives OneSignal in its Stack trace. But it does tell about Disk Readings on UIThread.. and when i enabled StrictMode it gives me an error on initialisation, where it calls getGcmPreferences and tries to get sharedPreferences data inside this.

So as you said Disk readings should be moved off the main thread. So it would be great if this can be done on priority as my app is being de prioritised because of ANR's in our App. So its kind of a critical thing for us. Thanks alot

**Comment by shaniamjad** [comment]
@jkasten2 Can you tell me when this fix will be available? When would i be able to integrate this Fixed version on my game?
cc- @Alamgir

**Comment by jkasten2** [comment]
@shaniamjad This fix was included in [OneSignal SDK 3.6.3](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/3.6.3)
